src/ContentValidationListener.php

```diff
@@ -356,6 +365,64 @@ protected function useRawData($controllerService)
         return false;
     }
 
+    /**
+     * @param $controllerService
+     * @return bool
+     */
+    protected function shouldRemoveEmptyData($controllerService)
+    {
+        if (! isset($this->config[$controllerService]['remove_empty_data'])
+            || (isset($this->config[$controllerService]['remove_empty_data'])
+                && $this->config[$controllerService]['remove_empty_data'] === true)
+        ) {
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * @param array $data
+     * @return array
+     */
+    protected function removeEmptyData(array $data = [])
+    {
+        $removeNull = function ($value) {
+            return ! is_null($value);
+        };
+
+        $data = array_filter($data, $removeNull);
+
+        if (empty($data)) {
+            return $data;
+        }
+
+        foreach ($data as $key => $value) {
+            if (is_array($value)) {
+                if (empty(array_filter($value, $removeNull))) {
+                    unset($data[$key]);
+                } else {
+                    $tmpValue = $this->removeEmptyData($value);
+
+                    // Additional check to ensure it's not an empty recursive result
+                    if (empty(array_filter($tmpValue, $removeNull))) {
+                        unset($data[$key]);
+                    } else {
+                        $data[$key] = $tmpValue;
+                    }
+
+                    // Destroy var so it's not accidentally used on a next iteration
+                    unset($tmpValue);
+                }
+            } else {
+                if (empty($value)) {
+                    unset($data[$key]);
+                }
+            }
```

Comment:
This can be simplified a ton:

```php
foreach ($data as $key => $value) {
    if (! is_array($value) && ! empty($value)) {
        continue;
    }
    if (! is_array($value)) {
        unset($data[$key]);
        continue;
    }
    if (empty(array_filter($value, $removeNull))) {
        unset($data[$key]);
        continue;
    }
    $tmpValue = $this->removeEmptyData($value);
    if (empty(array_filter($tmpValue, $removeNull))) {
        unset($data[$key]);
        continue;
    }
    $data[$key] = $tmpValue;
}
```

You don't need to unset `$tmpValue` as it gets re-assigned when the appropriate conditional occurs. By flattening the structure, we can better understand the various paths through the loop, which will make testing easier.

Reply:
Ah thank you, will have a look at it again soon. Atm pushing deadlines, so might be a wee while.

Reply:
Done